### PR TITLE
node agent: health check module

### PR DESCRIPTION
**Pull request by srinibas-zededa**
Signed-off-by: Srinibas Maharana <srinibas@zededa.com>

For detail spec and event-flow, please refer,
https://wiki.lfedge.org/display/EVE/NodeAgent%3A++EVE+node+health+monitor+module

### CODE REVIEW

File: pkg/pillar/cmd/nodeagent/nodeagent.go [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+	"github.com/lf-edge/eve/pkg/pillar/pidfile"
+	"github.com/lf-edge/eve/pkg/pillar/pubsub"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	"github.com/lf-edge/eve/pkg/pillar/zboot"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	agentName             = "nodeagent"
+	maxConfigGetFailCount = 5
```
Comment (eriknordmark) [inline pkg/pillar/cmd/nodeagent/nodeagent.go]: How is this useful? I we configure the get timer to every 5 seconds, this means a mere 25 seconds.

Is there a reason the time limit doesn't work??

File: pkg/pillar/cmd/nodeagent/nodeagent.go [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+)
+
+// Version : module version
+var Version = "No version specified"
+var globalConfig = types.GlobalConfigDefaults
+var rebootDelay = 30 // take a 30 second delay
+
+// XXX globals filled in by subscription handlers and read by handlemetrics
+// XXX could alternatively access sub object when adding them.
+var clientMetrics interface{}
```
Comment (eriknordmark) [inline pkg/pillar/cmd/nodeagent/nodeagent.go]: What does nodeagent do with these metrics?

File: pkg/pillar/cmd/nodeagent/nodeagent.go [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+var logmanagerMetrics interface{}
+var downloaderMetrics interface{}
+var networkMetrics types.NetworkMetrics
+
+type nodeagentContext struct {
```
Comment (eriknordmark) [inline pkg/pillar/cmd/nodeagent/nodeagent.go]: Isn't most of this dead stuff copied from zedagent? Need to prune to what is actually used in nodeagent.

File: pkg/pillar/cmd/baseosmgr/handlebaseos.go [hunk 4]
```diff
@@ -966,9 +966,9 @@ func baseOsSetPartitionInfoInStatus(ctx *baseOsMgrContext, status *types.BaseOsS
 	if partStatus != nil {
 		log.Infof("baseOsSetPartitionInfoInStatus(%s) %s found %+v\n",
 			status.Key(), partName, partStatus)
-		status.PartitionLabel = partName
-		status.PartitionState = partStatus.PartitionState
-		status.PartitionDevice = partStatus.PartitionDevname
+		status.PartitionLabel = strings.TrimSpace(partName)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/baseosmgr/handlebaseos.go]: Why did you need to add these here? Any why not add them in 10 other places?
If there is some input which can have spaces we should do a trimpspace where we use that input to set PartionState etc, not in the middle of the code.
If you want to assert that there is no leading or trailing space when you use it that would be fine, but not to actually remove them.

File: pkg/pillar/cmd/nodeagent/nodeagent.go (hunk 1 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/nodeagent/nodeagent.go]: I am seeing some getconfig request fails in between. To avoid the transient get request fail, I have added some default count, until we reset the upgrade validation test function, and start the test again.

File: pkg/pillar/cmd/nodeagent/nodeagent.go (hunk 2 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/nodeagent/nodeagent.go]: will remove these.

File: pkg/pillar/cmd/nodeagent/nodeagent.go (hunk 3 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/nodeagent/nodeagent.go]: sure, will remove these.

File: pkg/pillar/cmd/baseosmgr/handlebaseos.go (hunk 4 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/baseosmgr/handlebaseos.go]: sure, will remove these.

File: pkg/pillar/cmd/nodeagent/nodeagent.go (hunk 1 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/nodeagent/nodeagent.go]: But you'll fail after 25 seconds AFAICT even if the timer is set to be a lot higher.

Comment (srinibas-zededa) [reply pkg/pillar/cmd/nodeagent/nodeagent.go]: The controller get config request fails, periodically.. has a pattern of some temporary failure, and the next  config get request suceeds. In the 25 seconds, it would have recovered, and we will complete the upgrade validation test. 

{"file":"/pillar/zedcloud/send.go:103","func":"github.com/lf-edge/eve/pkg/pillar/zedcloud.SendOnAllIntf","level":"error","msg":"All attempts to connect to zedcloud.alpha.zededa.net/api/v1/edgedevice/config failed: [All attempts to connect to https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config using intf eth0 failed: [Get https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config: dial tcp: lookup zedcloud.alpha.zededa.net on 192.168.0.1:53: read udp 192.168.0.103:42256-\u003e192.168.0.1:53: read: connection refused] No management interfaces]","time":"2019-10-01T15:54:37.616136477Z"}
{"file":"/pillar/zedcloud/send.go:103","func":"github.com/lf-edge/eve/pkg/pillar/zedcloud.SendOnAllIntf","level":"error","msg":"All attempts to connect to zedcloud.alpha.zededa.net/api/v1/edgedevice/config failed: [All attempts to connect to https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config using intf eth0 failed: [Get https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config: dial tcp: lookup zedcloud.alpha.zededa.net on 192.168.0.1:53: read udp 192.168.0.103:39846-\u003e192.168.0.1:53: i/o timeout] No management interfaces]","time":"2019-10-01T15:54:42.620100823Z"}
{"file":"/pillar/zedcloud/send.go:103","func":"github.com/lf-edge/eve/pkg/pillar/zedcloud.SendOnAllIntf","level":"error","msg":"All attempts to connect to zedcloud.alpha.zededa.net/api/v1/edgedevice/config failed: [All attempts to connect to https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config using intf eth0 failed: [Get https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config: dial tcp: lookup zedcloud.alpha.zededa.net on 192.168.0.1:53: read udp 192.168.0.103:55826-\u003e192.168.0.1:53: read: connection refused] No management interfaces]","time":"2019-10-01T15:54:42.629004753Z"}
{"file":"/pillar/zedcloud/send.go:103","func":"github.com/lf-edge/eve/pkg/pillar/zedcloud.SendOnAllIntf","level":"error","msg":"All attempts to connect to zedcloud.alpha.zededa.net/api/v1/edgedevice/config failed: [All attempts to connect to https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config using intf eth0 failed: [Get https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config: dial tcp: lookup zedcloud.alpha.zededa.net on 192.168.0.1:53: read udp 192.168.0.103:35562-\u003e192.168.0.1:53: read: connection refused] No management interfaces]","time":"2019-10-01T15:54:48.671905065Z"}
{"file":"/pillar/zedcloud/send.go:103","func":"github.com/lf-edge/eve/pkg/pillar/zedcloud.SendOnAllIntf","level":"error","msg":"All attempts to connect to zedcloud.alpha.zededa.net/api/v1/edgedevice/config failed: [All attempts to connect to https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config using intf eth0 failed: [Get https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config: dial tcp: lookup zedcloud.alpha.zededa.net on 192.168.0.1:53: read udp 192.168.0.103:54735-\u003e192.168.0.1:53: read: connection refused] No management interfaces]","time":"2019-10-01T15:54:48.680837667Z"}
{"file":"/pillar/zedcloud/send.go:103","func":"github.com/lf-edge/eve/pkg/pillar/zedcloud.SendOnAllIntf","level":"error","msg":"All attempts to connect to zedcloud.alpha.zededa.net/api/v1/edgedevice/config failed: [All attempts to connect to https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config using intf eth0 failed: [Get https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config: dial tcp: lookup zedcloud.alpha.zededa.net on 192.168.0.1:53: read udp 192.168.0.103:59296-\u003e192.168.0.1:53: i/o timeout] No management interfaces]","time":"2019-10-01T15:54:55.255563556Z"}
{"file":"/pillar/zedcloud/send.go:103","func":"github.com/lf-edge/eve/pkg/pillar/zedcloud.SendOnAllIntf","level":"error","msg":"All attempts to connect to zedcloud.alpha.zededa.net/api/v1/edgedevice/config failed: [All attempts to connect to https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config using intf eth0 failed: [Get https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config: dial tcp: lookup zedcloud.alpha.zededa.net on 192.168.0.1:53: read udp 192.168.0.103:53224-\u003e192.168.0.1:53: read: connection refused] No management interfaces]","time":"2019-10-01T15:54:55.276441046Z"}
{"file":"/pillar/zedcloud/send.go:103","func":"github.com/lf-edge/eve/pkg/pillar/zedcloud.SendOnAllIntf","level":"error","msg":"All attempts to connect to zedcloud.alpha.zededa.net/api/v1/edgedevice/config failed: [All attempts to connect to https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config using intf eth0 failed: [Get https://zedcloud.alpha.zededa.net/api/v1/edgedevice/config: dial tcp: lookup zedcloud.alpha.zededa.net on 192.168.0.1:53: read udp 192.168.0.103:44054-\u003e192.168.0.1:53: i/o timeout] No management interfaces]","time":"2019-10-01T16:31:40.852548504Z"}

Comment (eriknordmark) [reply pkg/pillar/cmd/nodeagent/nodeagent.go]: That's a bug in the DNS handling in nim/devicenetwork which I just merged a fix into master.

But it isn't a count which matters; it is the 600 second timeout which matters. We shouldn't have any hardcoded numbers like this.

File: pkg/pillar/cmd/nodeagent/handletimers.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+func execReboot(ctx *nodeagentContext, rebootStr string) {
+
+	log.Infof("execReboot(): Reboot reason(%s)\n", rebootStr)
+	if ctx.needsReboot {
+		log.Infof("reboot flag is already set\n")
+		return
+	}
+
+	ctx.needsReboot = true
+	ctx.rebootReason = rebootStr
```
Comment (eriknordmark) [inline pkg/pillar/cmd/nodeagent/handletimers.go]: Where do you write the reboot reason to disk?
zedagent currently calls
			agentlog.RebootReason(errStr)
and you need the same AFAICT

File: pkg/pillar/cmd/nodeagent/handletimers.go [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+	publishNodeAgentStatus(ctx)
+
+	// if zedagent is not alive, handle it here
```
Comment (eriknordmark) [inline pkg/pillar/cmd/nodeagent/handletimers.go]: Why not consistently handle the failure reboot in one place? Makes for less test cases and less bug opportunities.

File: pkg/pillar/scripts/device-steps.sh [hunk 7]
```diff
@@ -330,6 +333,9 @@ if ! pgrep ledmanager >/dev/null; then
     echo "$(date -Ins -u) Starting ledmanager"
     ledmanager &
 fi
+sleep 10
```
Comment (eriknordmark) [inline pkg/pillar/scripts/device-steps.sh]: No need for random sleep. If you had to insert this, then your code has some race conditions which need to be fixed.

File: pkg/pillar/types/zedagenttypes.go [hunk 8]
```diff
@@ -207,3 +207,33 @@ type DatastoreConfig struct {
...
+// Key :
+func (status NodeAgentStatus) Key() string {
+	return status.Name
+}
+
+// ZedAgentStatus :
+type ZedAgentStatus struct {
+	Name                   string
+	LastConfigReceivedTime time.Time
+	ConfigGetFailCount     int
```
Comment (eriknordmark) [inline pkg/pillar/types/zedagenttypes.go]: How is this useful? The last receive is what you want to look at, and that should suffice.

File: pkg/pillar/cmd/zedagent/zedagent.go [hunk 9]
```diff
@@ -1232,6 +1252,50 @@ func handleZbootStatusDelete(ctxArg interface{}, key string,
...
+		return
+	}
+	updateInprogress := getconfigCtx.updateInprogress
+	ctx := getconfigCtx.zedagentCtx
+	ctx.remainingTestTime = status.RemainingTestTime
+	getconfigCtx.updateInprogress = status.UpdateInprogress
+	if status.NeedsReboot {
+		initiateDeviceReboot(ctx, status.RebootReason)
+	} else {
+		// if config reboot command was initiated and
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/zedagent.go]: There is also a case of an update which is deferred due to the previous update still being tested. Where do you handle that?

File: pkg/pillar/cmd/zedagent/zedagent.go [hunk 10]
```diff
@@ -878,6 +900,7 @@ func Run() {
 
 		case <-stillRunning.C:
 		}
+		doDeviceReboot(&zedagentCtx)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/zedagent.go]: I suspect the name of this function is misleading, unless you reboot the first time you go around the for loop.

File: pkg/pillar/cmd/nodeagent/nodeagent.go [hunk 11]
```diff
@@ -0,0 +1,484 @@
...
+	ctx := ctxArg.(*nodeagentContext)
+	if status := lookupZbootStatus(ctx, key); status == nil {
+		log.Infof("handleZbootStatusDelete: unknown %s\n", key)
+		return
+	}
+	log.Infof("handleZbootStatusDelete(%s) done\n", key)
+}
+
+// monitor device led blink change events
+func handleLedBlinkConfigModify(ctxArg interface{},
```
Comment (eriknordmark) [inline pkg/pillar/cmd/nodeagent/nodeagent.go]: The LED blinking was never designed to be operationally robust; merely to give feedback to an installer during onboard, So you shouldn't use it to guage the state of the system.

Furthermore, the logic here seems to have more than 10 cases. I think this needs to be explained in a state transition diagram, because there should only be a few cases:
1. Special onboarding - wait forever if we get 404 errors; if we don't get 404 errors then we should presumably reboot after ResetIfCloudGoneTime amount of time (to handle a hung device driver). The latter is not something we have today.
2. If baseimage update in progress, then if we don't get a config in the Fallback time, then reboot.
3. If not baseimage update in progress, then if we don't get a config in the ResetIfCloudGoneTime, then reboot.
Plus the logic for testcomplete. 
But that is an order of magnitude simpler than all of the different if cases in this function.

File: pkg/pillar/cmd/nodeagent/handlebaseos.go [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+// mark the zedcloud health/connectivity test complete flag
+// for baseosmgr to pick up and complete the partition activation
+func initiateBaseOsZedCloudTestComplete(ctx *nodeagentContext) {
+	if !ctx.updateInprogress {
+		return
+	}
+	log.Infof("initiateBaseOsZedCloudTestComplete(%s)\n", ctx.curPart)
+	// get the current partition zboot config and status
+	zbootConfig := lookupZbootConfig(ctx, ctx.curPart)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/nodeagent/handlebaseos.go]: Initially isn't zbootConfig absent? Where to do initially create it for the two partitions?

File: pkg/pillar/cmd/nodeagent/handlebaseos.go [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+		return
+	}
+	// get the baseos status information, for current partition
+	status := lookupBaseOsStatusOnPartition(ctx, ctx.curPart)
+	if status == nil {
+		log.Errorf("baseOs(%s) status get fail\n", ctx.curPart)
+		return
+	}
+	// the installed and configured version do not match
+	if status.BaseOsVersion != zbootStatus.ShortVersion {
```
Comment (eriknordmark) [inline pkg/pillar/cmd/nodeagent/handlebaseos.go]: This seems like an opportunity for a device to become un-updatable, because when we don't signal TestComplete to baseosmgr to clear the inprogress bit, I don't think we can update to another baseos image. At best we can recover by rebooting.

So ... why do you need to do this check here?

File: pkg/pillar/cast/cast.go [hunk 14]
```diff
@@ -401,7 +401,7 @@ func CastUuidToNum(in interface{}) types.UuidToNum {
 	return output
 }
 
-// ZbootConfig converts/casts from an interface to a ZbootConfig
+// ZbootConfig : casts interface to ZbootConfig
```
Comment (eriknordmark) [inline pkg/pillar/cast/cast.go]: I note that ZbootConfig is only about TestComplete. At some point in time it might make sense to rename it for clarity.

File: pkg/pillar/cmd/zedagent/parseconfig.go [hunk 15]
```diff
@@ -2153,15 +2135,13 @@ func handleReboot(getconfigCtx *getconfigContext) {
...
-
-	log.Infof("startExecReboot: scheduling exec reboot\n")
-
-	//timer was started, stop now
+// Used by doDeviceReboot only
+func startExecReboot(reasonStr string) {
+	//timer has already been started, return
 	if rebootTimer != nil {
-		rebootTimer.Stop()
+		return
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/parseconfig.go]: Why this change?

File: pkg/pillar/cmd/zedagent/zedagent.go [hunk 16]
```diff
@@ -538,11 +558,15 @@ func Run() {
 			agentlog.RebootReason(errStr)
 			execReboot(true)
 			agentlog.CheckMaxTime(agentName, start)
+
+		case <-stillRunning.C:
+			doDeviceReboot(&zedagentCtx)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/zedagent.go]: This is odd. stillRunning is about the touch file. Do you need that? If not pick a different timer name. But see other comment about deviceReboot boolean.

File: pkg/pillar/cmd/nodeagent/nodeagent.go [hunk 17]
```diff
@@ -0,0 +1,484 @@
...
+	log.Infof("Waiting for GCInitialized\n")
+	for !nodeagentCtx.GCInitialized {
+		select {
+		case change := <-subGlobalConfig.C:
+			subGlobalConfig.ProcessChange(change)
+
+		case change := <-subLedBlinkConfig.C:
+			subLedBlinkConfig.ProcessChange(change)
+
+		case <-stillRunning.C:
```
Comment (eriknordmark) [inline pkg/pillar/cmd/nodeagent/nodeagent.go]: Please use the new pattern where StillRunning(agentName) is called after the select to handle the lack of fairness in the random select.

File: pkg/pillar/cmd/nodeagent/nodeagent.go [hunk 18]
```diff
@@ -0,0 +1,484 @@
...
+	for !nodeagentCtx.deviceRegistered {
+		select {
+		case change := <-subGlobalConfig.C:
+			subGlobalConfig.ProcessChange(change)
+
+		case change := <-subLedBlinkConfig.C:
+			subLedBlinkConfig.ProcessChange(change)
+
+		case <-stillRunning.C:
+			agentlog.StillRunning(agentName)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/nodeagent/nodeagent.go]: Ditto

File: pkg/pillar/cmd/nodeagent/handletimers.go [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	"github.com/lf-edge/eve/pkg/pillar/zboot"
+	log "github.com/sirupsen/logrus"
+)
+
+// node health timer funcions
+
+// called periodically from agentwatch context
+func handleDeviceTimers(ctx *nodeagentContext) {
```
Comment (eriknordmark) [inline pkg/pillar/cmd/nodeagent/handletimers.go]: Why did you invent a new pattern where you periodically check? I think the explicit t1 and t2 timers in zedagent is easier to understand instead of depending on the stillRunning timer.

File: pkg/pillar/cmd/nodeagent/nodeagent.go [hunk 20]
```diff
@@ -0,0 +1,484 @@
...
+
+	// baseline timers, again
+	nodeagentCtx.lastConfigReceivedTime = time.Now()
+
+	log.Infof("Waiting for device registration check\n")
+	for !nodeagentCtx.deviceRegistered {
+		select {
+		case change := <-subGlobalConfig.C:
+			subGlobalConfig.ProcessChange(change)
+
```
Comment (eriknordmark) [inline pkg/pillar/cmd/nodeagent/nodeagent.go]: Please insert the new check for long time following the pattern in the other agents for all of the cases in the select statements.

File: pkg/pillar/cmd/nodeagent/nodeagent.go [hunk 21]
```diff
@@ -0,0 +1,484 @@
...
+	// come up in this time period,
+	//
+	// On zedbox modules activation, nodeagent will
+	// track the zedcloud connectivity events
+	//
+	// These timer functions will be tracked using
+	// led blinker and cloud connectionnectivity status.
+
+	// baseline timers, again
+	nodeagentCtx.lastConfigReceivedTime = time.Now()
```
Comment (eriknordmark) [inline pkg/pillar/cmd/nodeagent/nodeagent.go]: I suspect you should leave this at zero and then check if it has been set by zedagent.
Note that deviceagent starts before ntp, hence your time could be the year 1970 or 2016.
Recording a timestamp would fail on device which doesn't have a hardware clock.

File: pkg/pillar/cmd/zedagent/handlebaseos.go [hunk 22]
```diff
@@ -78,74 +60,25 @@ func lookupZbootStatus(ctx *zedagentContext, key string) *types.ZbootStatus {
...
-
-		if ctx.rebootCmdDeferred {
-			log.Infof("TestComplete and deferred reboot\n")
-			ctx.rebootCmdDeferred = false
-			duration := time.Second * time.Duration(rebootDelay)
-			rebootTimer = time.NewTimer(duration)
-			go handleReboot(ctx.getconfigCtx)
-		}
+// on baseos install and activate set, the device reboot is initiated
+func initiateDeviceReboot(ctx *zedagentContext, infoStr string) {
```
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/handlebaseos.go]: This pattern of a boolean to defer work is odd.
Can you change the ctx,deviceReboot to be a channel which you send on here, and the select statements receive on it? Such an approach is more maintainable IMHO.

File: pkg/pillar/cmd/nodeagent/nodeagent.go [hunk 23]
```diff
@@ -0,0 +1,484 @@
...
+			agentlog.StillRunning(agentName)
+			handleDeviceTimers(&nodeagentCtx)
+		}
+		if isZedAgentAlive(&nodeagentCtx) {
+			nodeagentCtx.deviceRegistered = true
+		}
+	}
+
+	// take a time out, for zedbox modules activation
+	time.Sleep(10)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/nodeagent/nodeagent.go]: If you need this then you have some race condition in the code.

File: pkg/pillar/cmd/nodeagent/nodeagent.go [hunk 24]
```diff
@@ -0,0 +1,484 @@
...
+
+	// rebase timers, again
+	nodeagentCtx.lastConfigReceivedTime = time.Now()
```
Comment (eriknordmark) [inline pkg/pillar/cmd/nodeagent/nodeagent.go]: Shouldn't you get this from zedagent?

File: pkg/pillar/cmd/nodeagent/handletimers.go [hunk 25]
```diff
@@ -0,0 +1,188 @@
...
+		time.Duration(globalConfig.MintimeUpdateSuccess)
+	ctx.remainingTestTime = successLimit
+	ctx.testInprogress = true
+}
+
+func resetTestStartTime(ctx *nodeagentContext) {
+	if !ctx.testInprogress {
+		return
+	}
+	// the led status has changed, incase, the test is in progress,
```
Comment (eriknordmark) [inline pkg/pillar/cmd/nodeagent/handletimers.go]: As long as you've moved past the onboarding case, why do you need to reset things?
Potentially you might want to wait (for 5 minutes like waitforaddr.go) for DeviceNetworkStatus indicating that there are usable addresses, but after that the Fallback time should apply.

FWIW the waitforaddr.go should really be a package which we'd use in the startup code for all the agents, so that the agents can start independently. But that is a subsequent simplification of device-steps.sh.

File: pkg/pillar/cmd/nodeagent/handletimers.go [hunk 26]
```diff
@@ -0,0 +1,188 @@
...
+	executeConfigGetFailState(ctx)
+}
+
+func updateLastConfigReceivedTime(ctx *nodeagentContext,
+	status types.ZedAgentStatus) {
+	log.Debugf("lastConfigReceivedtime: %v, failCount: %d \n",
+		status.LastConfigReceivedTime, status.ConfigGetFailCount)
+	// time stamp has not changed, since last config receive
+	if ctx.lastConfigReceivedTime == status.LastConfigReceivedTime {
+		handleConfigGetFail(ctx)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/nodeagent/handletimers.go]: How is a lack of a change a failure? You don't know whether you might get false notifications from pubsub or whether some other field in the published data has changed.

File: pkg/pillar/cmd/nodeagent/handletimers.go [hunk 27]
```diff
@@ -0,0 +1,188 @@
...
+func setTestStartTime(ctx *nodeagentContext) {
+	// test has already started or, the current
+	// partition is not in inprogress state,
+	// nothing to do
+	if !ctx.updateInprogress || ctx.testInprogress {
+		return
+	}
+	// only when current partition state is in progress
+	// start the test
+	ctx.upgradeTestStartTime = time.Now()
```
Comment (eriknordmark) [inline pkg/pillar/cmd/nodeagent/handletimers.go]: This can be problematic if the hardware doesn't have a time of day chip + battery since we don't run ntp until after we've started nodeagent. Using a normal timer as in zedagent is robust against the ToD changing.

File: pkg/pillar/cmd/nodeagent/nodeagent.go (hunk 1 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/nodeagent/nodeagent.go]: Sure, have made the changes, and also removed the ledblinker config for deciding the device registration state.

File: pkg/pillar/cmd/nodeagent/handletimers.go (hunk 5 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/nodeagent/handletimers.go]: While publishing the nodeagent status, it is picked up by the zedagent, nodeagent status handler and pushed as part of device reboot string.

File: pkg/pillar/types/zedagenttypes.go [hunk 28]
```diff
@@ -207,3 +207,40 @@ type DatastoreConfig struct {
...
+func (status NodeAgentStatus) Key() string {
+	return status.Name
+}
+
+type ConfigGetStatus uint8
+
+const (
+	CONFIG_GET_SUCCESS ConfigGetStatus = iota + 1
+	CONFIG_GET_FAIL
+	CONFIG_GET_FAIL_400
```
Comment (eriknordmark) [inline pkg/pillar/types/zedagenttypes.go]: Now or later we should add a CONFIG_GET_READ_SAVED to indicate that the config exists, but is from a checkpoint. That is useful to start the GC in other agents.

File: pkg/pillar/cmd/nodeagent/handletimers.go (hunk 5 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/nodeagent/handletimers.go]: Yes, in case zedagent is running. But nodeagent also handles the case when  only client is running, and I assumed that is why you have code where you have nodeagent do the reboot itself. In that case it needs to write the reason.

File: pkg/pillar/cmd/nodeagent/nodeagent.go (hunk 24 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/nodeagent/nodeagent.go]: sure

File: pkg/pillar/cmd/nodeagent/nodeagent.go (hunk 23 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/nodeagent/nodeagent.go]: removed

File: pkg/pillar/cmd/zedagent/handlebaseos.go (hunk 22 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/zedagent/handlebaseos.go]: have moved the functionality to  a channel/reboot handler task primitive.

File: pkg/pillar/cmd/nodeagent/handletimers.go (hunk 19 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/nodeagent/handletimers.go]: sure

File: pkg/pillar/cmd/nodeagent/nodeagent.go (hunk 18 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/nodeagent/nodeagent.go]: sure

File: pkg/pillar/cmd/nodeagent/nodeagent.go (hunk 17 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/nodeagent/nodeagent.go]: sure

File: pkg/pillar/cmd/zedagent/zedagent.go (hunk 16 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/zedagent/zedagent.go]: sure

File: pkg/pillar/cmd/zedagent/parseconfig.go (hunk 15 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/zedagent/parseconfig.go]: No point of resetting this for device state machine induced reboot events. They are bound to get executed. Resetting rebootTimer makes sense for opCmd based reboot commands.

File: pkg/pillar/cmd/nodeagent/handlebaseos.go (hunk 13 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/nodeagent/handlebaseos.go]: removed the check.

File: pkg/pillar/cmd/nodeagent/handlebaseos.go (hunk 12 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/nodeagent/handlebaseos.go]: the zboot configs are published inside nodeagent .Run() function.

File: pkg/pillar/cmd/nodeagent/nodeagent.go (hunk 11 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/nodeagent/nodeagent.go]: sure, removed LED blinker based handling.

File: pkg/pillar/cmd/zedagent/zedagent.go (hunk 10 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/zedagent/zedagent.go]: sure

File: pkg/pillar/cmd/zedagent/zedagent.go (hunk 9 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/zedagent/zedagent.go]: We handle this case, in this function itself.

File: pkg/pillar/types/zedagenttypes.go (hunk 8 above)
Comment (srinibas-zededa) [reply pkg/pillar/types/zedagenttypes.go]: removed.

File: pkg/pillar/scripts/device-steps.sh (hunk 7 above)
Comment (srinibas-zededa) [reply pkg/pillar/scripts/device-steps.sh]: removed

File: pkg/pillar/cmd/nodeagent/handletimers.go (hunk 6 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/nodeagent/handletimers.go]: When the zedagent is active, the idea is to gracefully shutdown the application instances, as zedagent is the publisher of application instance config. Otherwise, it will get executed in the nodeagent context itself.

File: pkg/pillar/cmd/nodeagent/handletimers.go (hunk 5 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/nodeagent/handletimers.go]: Sure, have added them

File: pkg/pillar/types/zedagenttypes.go (hunk 28 above)
Comment (srinibas-zededa) [reply pkg/pillar/types/zedagenttypes.go]: Have added, the enum and handling inside zedagent code, for  ConfigGetReadSaved status.

File: pkg/pillar/cmd/nodeagent/handletimers.go (hunk 27 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/nodeagent/handletimers.go]: Have simplified the time function. It is not going to use ToD value.

File: pkg/pillar/cmd/nodeagent/handletimers.go (hunk 25 above)
Comment (srinibas-zededa) [reply pkg/pillar/cmd/nodeagent/handletimers.go]: Have added the network connectivity time out handling.

**Comment by srinibas-zededa** [comment]
> Hey @srinibas-zededa I've just noticed that README.md wasn't updated to include vaultmgr in the list of microservices. Can you please do that?

Sure done.

**Comment by srinibas-zededa** [comment]
Sure, Have added the details for vaultmgr and nodeagent to README.md
